### PR TITLE
Adding /plan endpoint

### PR DESCRIPTION
**Pull request by CuijieLu**

### CODE REVIEW

File: backend/controllers/RunsController.js [hunk 1]
```diff
@@ -49,7 +51,29 @@ const columns = [
...
+  
+      var sampleList = createSampleList(sampleListTest);
+  
+      var runTypeGroup = groupSampleByRunType(sampleList);
+      for (const key of runTypeGroup.keys()){
+          result1.set(key, getCollisionGroup(runTypeGroup.get(key)));
+      }
+  
+      // Return result
+      return apiResponse.successResponseWithData(res, 'success', Object.fromEntries(result1));
```
Comment (DavidStreid) [inline backend/controllers/RunsController.js:73]: maybe a note about why `Object.fromEntries(result1)` is used b/c `Map` can't be serialized, or use an object from the start, e.g.
```
let sampleMap = {};
...
sampleMap[key] = getCollisionGroup(runTypeGroup.get(key));
...
return apiResponse.successResponseWithData(res, 'success', sampleMap);
```

File: backend/controllers/RunsController.js [hunk 2]
```diff
@@ -60,11 +84,12 @@ const columns = [
 exports.getRuns = [
   // authenticateRequest,
   function (req, res) {
-    logger.log('info', 'Retrieving random quote');
     let key = 'RUNS';
     let retrievalFunction = () => getRuns();
 
-    
+
+    logger.info('info', 'doing other work');
```
Comment (DavidStreid) [inline backend/controllers/RunsController.js:91]: sorry for adding this - please remove

File: backend/helpers/classes/Sample.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
+const { logger } = require('../winston');
+
+// define class of individual samples
+
+class sample{
```
Comment (DavidStreid) [inline backend/helpers/classes/Sample.js:5]: Please capitalize - https://google.github.io/styleguide/javascriptguide.xml?showone=Naming#Naming
e.g.
```
class Sample {
    ...
}
```

File: backend/helpers/classes/Sample.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    constructor(poolID, sampleID, recipe, requestID, RequestName, sampleConc, units, volume, barcodeSeq, runLength, readsRequest, readsRemaining){
```
Comment (DavidStreid) [inline backend/helpers/classes/Sample.js:6]: Please add an `@param` for each w/ their type - 
e.g.
```
/**
 * @param {string} poolID Identifier of pool
 * @param {string} sampleID ...
    ...
 * @constructor
 */
```

File: backend/helpers/classes/Sample.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            }
+        }
+        return barcodeList;
+    }
+
+    filterSample(sample) {
+        return sample.runLength !== undefined;
+    }
+
+    constructor(listOfSamples){
```
Comment (DavidStreid) [inline backend/helpers/classes/Sample.js:64]: Please move to top of class & add `@param` strings

File: backend/helpers/classes/Sample.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        this.runLength = runLength;
+        this.readsRequest = Number(readsRequest);
+        this.readsRemaining = readsRemaining;
+    }
+
+}
+
+// class of pooled samples
+
+class pooledSample{
```
Comment (DavidStreid) [inline backend/helpers/classes/Sample.js:29]: capitalize

**Review by DavidStreid — COMMENTED** [review]
Looks good! Some notes on styling. Please follow this guide for documenting functions & classes - https://google.github.io/styleguide/javascriptguide.xml?showone=Comments#Comments

File: backend/controllers/RunsController.js [hunk 7]
```diff
@@ -49,7 +51,29 @@ const columns = [
...
 
 
+exports.plan = [
+  function(req, res){
 
+    getRuns()
+    .then((result)=>{
+      var result1 = new Map();
+      var sampleListTest = [];
+      for(let i = 0; i < result.data.length; i++){
```
Comment (apatruno618) [inline backend/controllers/RunsController.js:61]: Can you try using forEach()? I find it makes my code more readable
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

File: backend/controllers/RunsController.js [hunk 8]
```diff
@@ -49,7 +51,29 @@ const columns = [
 ];
 
 
+exports.plan = [
+  function(req, res){
 
+    getRuns()
```
Comment (apatruno618) [inline backend/controllers/RunsController.js:57]: I find the name of this function to be misleading given that the end goal of the plan endpoint is to plan a run, can we think of an alternative name? Maybe something like retrieveSamplesFromLimsQueue()?

File: backend/helpers/planRunsUtil.js [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+                return true;
+            }
+        }
+    }
+    return false;
+}
+//function for checking whether barcode collison happens between two samples
+function sampleBarcodeCollision(sample1, sample2, numOfMismatch){
+    if (sample1.poolID !== undefined && sample2.poolID !== undefined 
+    && sample1.containNormal == true && sample2.containNormal == true){
```
Comment (apatruno618) [inline backend/helpers/planRunsUtil.js:138]: you can simplify this to `&& sample1.containNormal && sample2.containNormal){`
Remember to use strict equality (===) in Javascript
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality

**Review by apatruno618 — COMMENTED** [review]
In general, please add more comments and use more descriptive variable/function names

File: backend/helpers/classes/Sample.js [hunk 10]
```diff
@@ -1,10 +1,25 @@
 const { logger } = require('../winston');
 
-// define class of individual samples
-
-class sample{
+/**
+ * Class that represents a sample object.
+ * 
+ * @param {string} poolID This will equal to undefined if it is an individual sample, information coming from pool
```
Comment (apatruno618) [inline backend/helpers/classes/Sample.js:6]: Here are some naming conventions you should follow. In general, use lowerCamelCase
https://google.github.io/styleguide/jsguide.html#naming

File: backend/helpers/classes/Sample.js [hunk 11]
```diff
@@ -13,29 +28,61 @@ class sample{
...
+ * Class that represents pooled samples which means samples that have poolID with it. 
+ *     Each PooledSample object will represent one pool.
+ * 
+ * @param {Array} listOfSamples An array of Sample objects.
+ * 
+ */
+class PooledSample{
+    constructor(listOfSamples){ 
+        this.readsRequest = this.getTotalReads(listOfSamples);
+        this.containNormal = this.ifContainNormal(listOfSamples);
```
Comment (apatruno618) [inline backend/helpers/classes/Sample.js:53]: This isn't checking that the group contains any normal sample, its checking if the group contains a Pooled Normal control, right?
Can you incorporate that into the language? For example,
`this.containsAPooledNormal = this.checkForPooledNormals(listOfSamples);`

**Review by apatruno618 — APPROVED** [review]

File: backend/controllers/RunsController.js (hunk 8 above)
Comment (DavidStreid) [reply backend/controllers/RunsController.js]: will save for later refactor pull request

**Review by DavidStreid — APPROVED** [review]
Nice!

**Pull request merged**